desc/objectives/nae_utils.py

```diff
@@ -446,3 +450,396 @@ def make_RZ_cons_1st_order(qsc, desc_eq, fix_lambda=False, N=None):
+        Zbasis_sin = FourierSeries(N=N, NFP=nfp, sym=False)
+
+    bases["Rbasis_cos"] = Rbasis
+    bases["Rbasis_sin"] = Rbasis_sin
+    bases["Zbasis_cos"] = Zbasis
+    bases["Zbasis_sin"] = Zbasis_sin
+
+    grid = LinearGrid(M=0, L=0, zeta=phi, NFP=nfp)
+    Rtrans = Transform(grid, Rbasis, build_pinv=True, method="auto")
+    Ztrans = Transform(grid, Zbasis, build_pinv=True, method="auto")
+    Rtrans_sin = Transform(grid, Rbasis_sin, build_pinv=True, method="auto")
+    Ztrans_sin = Transform(grid, Zbasis_sin, build_pinv=True, method="auto")
+
+    # if stell sym we should be able to figure out the phi sym for each term
+    # R cos terms need a cos phi basis
+    # R sin terms need sin phi basis
+    # Z cos terms need sin phi basis
+    # Z sin terms need cos phi basis
+
+    R_2_0_n = Rtrans.fit(R_2_0)
+    R_2_2_n = Rtrans.fit(R_2_2)
+    R_2_neg2_n = Rtrans_sin.fit(R_2_neg2)
+
+    coeffs["R_2_0_n"] = R_2_0_n
+    coeffs["R_2_2_n"] = R_2_2_n
+    coeffs["R_2_neg2_n"] = R_2_neg2_n
+
+    Z_2_0_n = Ztrans_sin.fit(Z_2_0)
+    Z_2_2_n = Ztrans_sin.fit(Z_2_2)
+    Z_2_neg2_n = Ztrans.fit(Z_2_neg2)
+
+    coeffs["Z_2_0_n"] = Z_2_0_n
+    coeffs["Z_2_2_n"] = Z_2_2_n
+    coeffs["Z_2_neg2_n"] = Z_2_neg2_n
+
+    return coeffs, bases
+
+
+def _calc_2nd_order_constraints(  # noqa: C901 - FIXME - simplify
+    qsc, desc_eq, coeffs, bases
+):
+    """Creates 2nd order NAE constraints for a DESC eq based off given qsc eq.
+
+    Parameters
+    ----------
+        qsc (Qsc): pyQsc Qsc object to use as the NAE constraints on the DESC eq
+        desc_eq (Equilibrium): desc equilibrium to constrain
+        coeffs: dict, dictionary of arrays with keys like 'X_L_M_n', where
+                X is R or Z, L is 1 or 2, and M is 0,1, or 2, are the
+                NAE Fourier (in tor. phi) coeffs of radial order L and poloidal order M
+        bases: dict, dictionary of Rbasis_cos, Rbasis_sin, Zbasis_cos, Zbasis_sin,
+            the FourierSeries basis objects used to obtain the coefficients, where
+            _cos or _sin denotes the symmetry of the Fourier series.
+            symmetry is such that the R or Z coefficients is stellarator symmetric
+            i.e. R_1_1_n uses the Rbasis_cos, since cos(theta)*cos(phi) is
+             stellarator symmetric for R i.e. R(-theta,-phi) = R(theta,phi)
+            and Z_1_1_n uses the Zbasis_sin as the term is cos(theta)*sin(phi)
+            since Z(-theta,-phi) = - Z(theta,phi) for Z stellarator symmetry
+
+    Returns
+    -------
+        Rconstraints (tuple): tuple of FixSumModesR constraints corresponding
+         to constraining the O(rho) DESC coefficients, to be used in
+         constraining a DESC equilibrium solve
+        Zconstraints (tuple): tuple of FixSumModesZ constraints corresponding
+         to constraining the O(rho) DESC coefficients, to be used in
+         constraining a DESC equilibrium solve
+    Notes
+    ----
+        follows eqns 30a and 30b in NAE2DESC document
+    """
```

Comment:
Do you have a test or something like that to check the correct implementation of these coefficients?

Reply:
None beyond testing a solve with them, I can make one to test at least like the symmetries of each one or something though.